packages/forms/src/documents/pdf/extract.ts

```diff
@@ -1,23 +1,57 @@
-import * as pdfLib from 'pdf-lib';
+import {
+  PDFDocument,
+  PDFName,
+  PDFDict,
+  PDFTextField,
+  PDFField,
+  PDFCheckBox,
+  PDFDropdown,
+  PDFOptionList,
+  PDFRadioGroup,
+} from 'pdf-lib';
 
 import { stringToBase64 } from '../util';
 import type { DocumentFieldValue, DocumentFieldMap } from '../types';
 
+// TODO: copied from pdf-lib acrofield internals, check if it's already exposed outside of acroform somewhere
+export const getWidgets = async (pdfDoc: PDFDocument): Promise<PDFDict[]> => {
+  return pdfDoc.context
+    .enumerateIndirectObjects()
+    .map(([, obj]) => obj)
+    .filter(
+      obj =>
+        obj instanceof PDFDict &&
+        obj.get(PDFName.of('Type')) === PDFName.of('Annot') &&
+        obj.get(PDFName.of('Subtype')) === PDFName.of('Widget')
+    )
+    .map(obj => obj as PDFDict);
+};
+
 export const getDocumentFieldData = async (
   pdfBytes: Uint8Array
 ): Promise<DocumentFieldMap> => {
-  const pdfDoc = await pdfLib.PDFDocument.load(pdfBytes);
+  const pdfDoc = await PDFDocument.load(pdfBytes);
+  const widgets = await getWidgets(pdfDoc);
+
+  pdfDoc.catalog.set(
+    PDFName.of('AcroForm'),
+    pdfDoc.context.obj({
+      Fields: widgets.map(widget => pdfDoc.context.getObjectRef(widget)), // array of widget refs
+    })
+  );
+
```

Comment:
This is where we add all of the widgets to the acroform, which will now get pulled into form via `pdfDoc.getForm()` below

Reply:
Interesting, so this is backfilling poorly annotated fields into the source PDF? That seems like an interesting approach. I like that it gives us a clear picture of the "error" of the source PDF, that we could perhaps inform the user about.